specs/Standards/StorageManagement.md

```diff
@@ -0,0 +1,268 @@
+
+Prior art:
+
+- A previous fungible token standard ([NEP-21](https://github.com/near/NEPs/pull/21)) highlighting how [storage was paid](https://github.com/near/near-sdk-rs/blob/1d3535bd131b68f97a216e643ad1cba19e16dddf/examples/fungible-token/src/lib.rs#L92-L113) for when increasing the allowance of an escrow system.
+
+## Guide-level explanation
+
+We should be able to do the following:
+
+- Have an account sign a transaction, attaching a deposit of Ⓝ that will cover *their own* storage for a smart contract.
+- Have an account sign a transaction, attaching a deposit of Ⓝ that will cover the storage for another account to use a smart contract.
+- Allow an account to request refunds in Ⓝ for excess storage deposits on a given contract.
+- Use a view function call to determine how much Ⓝ should be attached as a deposit in order for an account to "pay their share" in storage.
+
+### Example scenarios
+
+Imagine a social media smart contract which passes storage costs onto users for posts and follower data. Let's say this this contract is deployed at account `social`, and that the minimum storage required to interact with this contract is 0.1Ⓝ, of which 0.00235Ⓝ will be used to register the account and the rest will be kept to allow further storage-adding interactions with the contract.
+
+Let's follow a user, Alice with account `alice`, as she interacts with `social` through the following scenarios:
+
+1. Registration
+2. Registering a friend, Bob with account `bob`
+3. Attempting to take action which exceeds paid-for storage; increasing storage deposit
+4. Removing storage and reclaiming excess deposit
+
+#### 1. Account registers with `social`
+
+**High-level explanation**
+
+1. Alice checks if she is already registered with `social`.
+2. Alice determines how much NEAR is needed to register an account with the `social` contract.
+3. Alice issues a transaction to deposit Ⓝ for her account.
+
+**Technical calls**
+
+1. Alice queries a view-only method to determine if she already has storage on this contract with `social::storage_balance_of({"account_id": "alice"})`. Using [NEAR CLI](https://docs.near.org/docs/tools/near-cli) to make this view call, the command would be:
+
+       near view social storage_balance_of '{"account_id": "alice"}'
+
+   The response:
+
+       View call: social.storage_balance_of({"account_id": "alice"})
+       { total: '0', available: '0' }
+
+2. Alice uses [NEAR CLI](https://docs.near.org/docs/tools/near-cli) to make a view call.
+
+       near view social storage_minimum_balance
+
+   The response:
+
+       '100000000000000000000000'
+
+   This is 0.1Ⓝ specified in yoctoⓃ ([yocto](https://www.metricconversion.us/prefixes.htm) = 10<sup>-24</sup>).
+
+3. Alice deposits the proper amount in a transaction by calling `social::storage_deposit` with the attached deposit of '0.1'. Using NEAR CLI:
```

Comment:
The Storage Management standard's interplay with the Account Registration standard is not well-defined. The members of the Standards Working Group with whom I've spoken all have different ideas about how a Storage Manager contract should behave, with regards to Account Registration. An example scenario that generates many different opinions:

Given a contract that implements Storage Management at `contract`
And given a user `alice` who has not yet interacted with `contract`
When `alice` first interacts with `contract`
Does she need to first call `ar_register`, or can she call only `storage_deposit`?

Other points of disagreement:

* Is every Storage Management contract required to also implement all Account Registration functions?
* Does `storage_balance_of` display account registration fee as part of the `total`? 

Given that Storage Management is not needed for Fungible Token, I propose we split Storage Management into a separate Pull Request. This will allow us to finalize the core set of standards needed for Fungible Token with Metadata, while continuing to add clarity to the Storage Management standard.

Reply:
Sure, we can split the Storage Management standard into another pull request.